src/parser/java.rs

```diff
@@ -0,0 +1,456 @@
+/// Instantiate a [`ParserStack`] with the provided parsers.
+///
+/// Usage:
+/// ```ignore
+/// parser_stack!(T, parsers...)
+/// ```
+///
+/// Creates a static variable in the current scope with the name `PARSER_STACK`
+/// containing a stack of parsers that result in type `T`.
+///
+/// Using this and [`some`], one can form an analogue to an "alternative" parser combinator:
+/// ```ignore
+/// pub fn scope(node: TSNode<'_>, content: &[u8]) -> Option<Scope> {
+///     parser_stack!(Scope => scope_enter, scope_exit);
+///     some(&mut PARSER_STACK.iter(), |parser| parser(node, content))
+/// }
+/// ```
+macro_rules! parser_stack {
+    ($type:ty => $($args:expr),* $(,)*) => {
+        static PARSER_STACK: ParserStack<$type> = Lazy::new(|| vec![$($args,)*]);
+    };
+}
+
+/// Parse the node as a scope event.
+#[tracing::instrument(skip_all)]
+pub fn scope(node: TSNode<'_>, content: &[u8]) -> Option<Scope> {
+    parser_stack!(Scope => scope_enter, scope_exit);
+    some(&mut PARSER_STACK.iter(), |parser| parser(node, content))
+}
+
+#[tracing::instrument(skip_all)]
+fn scope_enter(node: TSNode<'_>, _: &[u8]) -> Option<Scope> {
+    ensure_node_kind!(node, NODE_KIND_OPEN_BRACE);
+    Scope::Enter(Location::from(node)).pipe(Some)
+}
+
+#[tracing::instrument(skip_all)]
+fn scope_exit(node: TSNode<'_>, _: &[u8]) -> Option<Scope> {
+    ensure_node_kind!(node, NODE_KIND_CLOSE_BRACE);
+    Scope::Exit(Location::from(node)).pipe(Some)
+}
+
+/// Attempt to parse the node as any supported symbol.
+#[tracing::instrument(skip_all)]
+pub fn symbol(node: TSNode<'_>, content: &[u8]) -> Option<Symbol<Node>> {
+    parser_stack!(
+        Symbol<Node> =>
+        invocation,
+        method,
+        field,
+        constructor,
+        class,
+        import,
+        package
+    );
+    some(&mut PARSER_STACK.iter(), |parser| parser(node, content))
+}
+
+#[tracing::instrument(skip_all)]
+fn package(node: TSNode<'_>, content: &[u8]) -> Option<Symbol<Node>> {
+    ensure_node_kind!(node, "package_declaration");
+
+    let mut ctx = traverse_statement(node, content);
+    let name = parse_some!(ctx => content, if_kind, "scoped_identifier");
+
+    symbol!(Node::new_package(name.inner), name.loc)
+}
+
+#[tracing::instrument(skip_all)]
+fn import(node: TSNode<'_>, content: &[u8]) -> Option<Symbol<Node>> {
+    ensure_node_kind!(node, "import_declaration");
+
+    let mut ctx = traverse_statement(node, content);
+    let name = parse_some!(ctx => content, if_kind, "scoped_identifier");
+    symbol!(Node::new_import(name.inner), name.loc)
+}
+
+#[tracing::instrument(skip_all)]
+fn class(node: TSNode<'_>, content: &[u8]) -> Option<Symbol<Node>> {
+    ensure_node_kind!(node, "class_declaration");
+
+    // Since visibility modifiers come first, but are optional,
+    // to parse these we need backtracking in the iterator parser.
+    // I don't have this built yet, so just assume all modifiers are public
+    // for now.
+    let visibility = Visibility::Public;
```

Comment:
[Question] These should not matter right? If a vulnerable package's function is invoked, we need not care if it is within public or private modifier.

Reply:
Yes, this only really matters to reduce the number of rows in the precomputed call graphs table, by allowing us to constrain it to only store public functions.